source/fuzz/transformation_propagate_instruction_down.h

```diff
@@ -0,0 +1,123 @@
+#include "source/fuzz/transformation.h"
+#include "source/fuzz/transformation_context.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace fuzz {
+
+class TransformationPropagateInstructionDown : public Transformation {
+ public:
+  explicit TransformationPropagateInstructionDown(
+      const protobufs::TransformationPropagateInstructionDown& message);
+
+  TransformationPropagateInstructionDown(
+      uint32_t block_id, uint32_t phi_fresh_id,
+      const std::map<uint32_t, uint32_t>& successor_id_to_fresh_id);
+
+  // - it should be possible to apply this transformation to |block_id| (see
+  //   IsApplicableToBlock method).
+  // - every acceptable successor of |block_id| (see GetAcceptableSuccessors
+  //   method)
+  //   must have an entry in the |successor_id_to_fresh_id| map.
+  // - all values in |successor_id_to_fresh_id| and |phi_fresh_id| must be
+  //   unique and fresh.
+  bool IsApplicable(
+      opt::IRContext* ir_context,
+      const TransformationContext& transformation_context) const override;
+
+  // - Adds a clone of the propagated instruction into every acceptable
+  //   successor of |block_id|.
+  // - Removes the original instruction.
+  // - Creates an OpPhi instruction if possible, that tries to group created
+  //   clones.
+  // - If the original instruction's id was irrelevant - marks created
+  //   instructions as irrelevant. Otherwise, marks the created instructions as
+  //   synonymous to each other.
+  void Apply(opt::IRContext* ir_context,
+             TransformationContext* transformation_context) const override;
+
+  protobufs::Transformation ToMessage() const override;
+
+  // Returns true if this transformation can be applied to the block with id
+  // |block_id|. Concretely, returns true iff:
+  // - |block_id| must be a result id of some reachable basic block in the
+  //   module.
+  // - the block must have an instruction to propagate (see
+  //   GetInstructionToPropagate method).
+  // - the block must have at least one acceptable successor (see
+  //   GetAcceptableSuccessors method).
+  // - none of the acceptable successors may have OpPhi instructions that use
+  //   the original instruction.
+  // - it should be possible to replace every use of the original instruction
+  //   with some of the propagated instructions (or an OpPhi if we can create
+  //   it - see CanAddOpPhiInstruction method).
+  static bool IsApplicableToBlock(opt::IRContext* ir_context,
+                                  uint32_t block_id);
+
+  // Returns ids of successors of |block_id|, that can be used to propagate an
+  // instruction into. Concretely, a successor block is acceptable if all
+  // dependencies of the propagated instruction dominate it. Note that this
+  // implies that an acceptable successor must be reachable in the CFG.
+  static std::unordered_set<uint32_t> GetAcceptableSuccessors(
+      opt::IRContext* ir_context, uint32_t block_id);
```

Comment:
I think these comments would be even more valuable if you could back them up by a couple of small examples to illustrate cases that are applicable/acceptable and cases that are not.

Reply:
Fixed in 69b41e9.